server/service/src/sync/file_synchroniser.rs

```diff
@@ -131,87 +126,92 @@ impl FileSynchroniser {
-                                retry_at: Some(retry_at),
-                                error: Some(format!("{:?}", err)),
-                                ..file.clone()
-                            })?;
-                        }
-
-                        return Err(err);
-                    }
-                };
-
-                // Update database to record the file has been uploaded
-                sync_file_repo.update_status(&SyncFileReferenceRow {
-                    uploaded_bytes: file.total_bytes, // We always upload the whole file in one go
-                    status: SyncFileStatus::Done,
-                    error: None,
-                    ..file.clone()
-                })?;
+        let Some(sync_file_reference) = file_references.first() else {
+            return Ok(0);
+        };
+
+        // update the database to say we're uploading the file
+        sync_file_repo.update_status(&SyncFileReferenceRow {
+            status: SyncFileStatus::InProgress,
+            ..sync_file_reference.clone()
+        })?;
+
+        let file_category = StaticFileCategory::SyncFile(
+            sync_file_reference.table_name.to_owned(),
+            sync_file_reference.record_id.to_owned(),
+        );
+
+        let file = self
+            .static_file_service
+            .find_file(&sync_file_reference.id, file_category)?
+            .ok_or(FileSyncError::FileNotFound(sync_file_reference.id.clone()))?;
+
+        let file_handle = std::fs::File::open(file.path.clone())?;
+
+        let upload_result = self
+            .sync_api_v6
+            .upload_file(&sync_file_reference, &file.name, file_handle)
+            .await;
+
+        let Err(error) = upload_result
+        // On Success
+        else {
+            sync_file_repo.update_status(&SyncFileReferenceRow {
+                uploaded_bytes: sync_file_reference.total_bytes, // We always upload the whole file in one go
+                status: SyncFileStatus::Done,
+                error: None,
+                ..sync_file_reference.clone()
+            })?;
+
+            return Ok(file_references.len());
+        };
+
+        // On Error
+
+        // Update database to record the file has failed to upload
+        let sync_file_ref_update = if sync_file_reference.retries >= MAX_UPLOAD_ATTEMPTS {
+            SyncFileReferenceRow {
+                status: SyncFileStatus::PermanentFailure,
+                ..sync_file_reference.clone()
             }
-            None => {
-                // No files to upload
+        } else {
+            // Calculate the next retry time
+
+            // if we get a 404 error it probably means the sync_file_reference hasn't been synced yet.
+            // So wait 1 minute before retrying
+            // Otherwise, do an exponential backoff
+            let retry_at = match error.source {
+                SyncApiErrorVariantV6::ParsedError(SyncParsedErrorV6::SyncFileNotFound(_)) => {
```

Comment:
Not a huge one, but this error is then logged by `file_sync_driver`... i.e. in remote site server logs you see:

![Screenshot 2024-04-24 at 10 37 54 AM](https://github.com/msupply-foundation/open-msupply/assets/55115239/99dbdd6a-e76a-467a-8dec-7a54493fe609)

Initially I got a bit confused (why can't the remote server find its own file??) Then realised this is actually the error from central server... Probably this will happen a lot - sync_file_reference not synced yet? Perhaps worth logging a clearer message here on remote server?

Reply:
ooh, a little bit concerned that response Could not be parsed, i've added to a follow up issue (since a few PR's are stacked from this one)

https://github.com/msupply-foundation/open-msupply/issues/3666

Reply:
Totally agree with this comment, would be good to have this cleaned up.
I'm slightly tempted to just create a dummy `sync file reference` and just allow uploading without syncing first?
Similar to what we do with temperture_log/temperature_breach records that rely on each other...
Now that we're authenticating properly I think that would be ok?